contracts/voting-verifier/src/execute.rs

```diff
@@ -88,16 +94,67 @@ pub fn verify_messages(
 }
 
 pub fn vote(
-    _deps: DepsMut,
-    _info: MessageInfo,
-    _poll_id: String,
-    _votes: Vec<bool>,
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    poll_id: PollID,
+    votes: Vec<bool>,
 ) -> Result<Response, ContractError> {
-    todo!()
+    let mut poll = POLLS
+        .may_load(deps.storage, poll_id)?
+        .ok_or(ContractError::PollNotFound {})?;
+
+    poll.cast_vote(env.block.height, &info.sender, votes)?;
+    POLLS.save(deps.storage, poll_id, &poll)?;
+
+    Ok(Response::new().add_event(
+        Voted {
+            poll_id,
+            voter: info.sender,
+        }
+        .into(),
+    ))
 }
 
-pub fn end_poll(_deps: DepsMut, _poll_id: String) -> Result<Response, ContractError> {
-    todo!()
+pub fn end_poll(deps: DepsMut, env: Env, poll_id: PollID) -> Result<Response, ContractError> {
+    let mut poll = POLLS
+        .may_load(deps.storage, poll_id)?
+        .ok_or(ContractError::PollNotFound {})?;
+
+    let poll_result = poll.tally(env.block.height)?;
+    POLLS.save(deps.storage, poll_id, &poll)?;
+    let messages = clear_pending_message(deps.storage, poll_id)?;
+    if messages.len() != poll_result.results.len() {
+        panic!(
+            "poll {} results and pending messages have different length",
+            poll_id
+        )
+    }
+
+    let messages = messages
+        .iter()
+        .zip(poll_result.results.iter())
+        .filter_map(|(message, verified)| match *verified {
+            true => Some(message),
+            false => None,
+        })
+        .collect::<Vec<&Message>>();
+
+    for message in messages {
+        if !is_message_verified(deps.as_ref(), message)? {
+            VERIFIED_MESSAGES.save(deps.storage, &message.id, message)?;
+        }
+    }
+
+    Ok(Response::new()
+        .add_event(
+            PollEnded {
+                poll_id: poll_result.poll_id,
+                results: &poll_result.results,
+            }
+            .into(),
+        )
+        .set_data(to_binary(&EndPollResponse { poll_result })?))
```

Comment:
I agree with Chris on that it's weird for the `PollEnded` event to take `results` in as a reference.

Reply:
sure, can clone